applications/TrilinosApplication/custom_strategies/convergencecriterias/trilinos_mixed_generic_criteria.h

```diff
@@ -0,0 +1,376 @@
+                }
+            }
+
+            auto global_solution_norms_vector = rModelPart.GetCommunicator().GetDataCommunicator().SumAll(solution_norms_vector);
+            auto global_increase_norms_vector = rModelPart.GetCommunicator().GetDataCommunicator().SumAll(increase_norms_vector);
+            auto global_dofs_count = rModelPart.GetCommunicator().GetDataCommunicator().SumAll(dofs_count);
+
+            const double zero_tol = 1.0e-12;
+            for(int i = 0; i < mVariableSize; i++) {
+                if (global_solution_norms_vector[i] < zero_tol) {
+                    global_solution_norms_vector[i] = 1.0;
+                }
+            }
+
+            std::vector<TDataType> var_ratio(mVariableSize, 0.0);
+            std::vector<TDataType> var_abs(mVariableSize, 0.0);
+            for(int i = 0; i < mVariableSize; i++) {
+                var_ratio[i] = std::sqrt(global_increase_norms_vector[i] / global_solution_norms_vector[i]);
+                var_abs[i] = std::sqrt(global_increase_norms_vector[i]) / static_cast<TDataType>(global_dofs_count[i]);
+            }
+
+            // Output convergence status
+            if (this->GetEchoLevel() > 0) {
+                std::ostringstream stringbuf;
+                stringbuf << "CONVERGENCE CHECK:\n";
+                for(int i = 0; i < mVariableSize; i++) {
+                    const auto r_var_data = mVariableDataVector[i];
+                    const int key_map = mLocalKeyMap[r_var_data->Key()];
+                    stringbuf << " " << r_var_data->Name() << " : ratio = " << var_ratio[key_map] << "; exp.ratio = " << mRatioToleranceVector[key_map] << " abs = " << var_abs[key_map] << " exp.abs = " << mAbsToleranceVector[key_map] << "\n";
+                }
+                KRATOS_INFO("") << stringbuf.str();
+            }
+
+            // Check convergence
+            int is_converged = 0;
+            for (int i = 0; i < mVariableSize; i++) {
+                const auto r_var_data = mVariableDataVector[i];
+                const int key_map = mLocalKeyMap[r_var_data->Key()];
+                is_converged += var_ratio[key_map] <= mRatioToleranceVector[key_map] || var_abs[key_map] <= mAbsToleranceVector[key_map];
+            }
+
+            if (is_converged) {
+                KRATOS_INFO_IF("", this->GetEchoLevel() > 0) << "*** CONVERGENCE IS ACHIEVED ***" << std::endl;
+                return true;
+            } else {
+                return false;
+            }
+        } else {
+            // Case in which all the DOFs are constrained!
+            return true;
+        }
+    }
+
+    ///@}
+    ///@name Access
+    ///@{
+
+
+    ///@}
+    ///@name Inquiry
+    ///@{
+
+
+    ///@}
+    ///@name Input and output
+    ///@{
+
+
+    ///@}
+private:
+    ///@name Static Member Variables
+    ///@{
+
+
+    ///@}
+    ///@name Member Variables
+    ///@{
+
+    const int mVariableSize;
+    const std::vector<std::shared_ptr<VariableData>> mVariableDataVector;
```

Comment:
i see no reason for having a shared_ptr here so unless there is a good reason i would use

i would use a VariableData* here,

Reply:
No, we already tryied that and does not work. The pointer becomes invalid because the result of `std::get<x>` only exist inside the lambda function. Unless someone is actually holding the variable it gets destroyed.

Reply:
This is strange. It would mean that you are making a copy of the variable...which is something we are not allowing

Reply:
VariableData is not as protected as Variable...

Reply:
> This is strange. It would mean that you are making a copy of the variable...which is something we are not allowing

We point to the `VariableData` (base class of `Variable`). I'd add that the same raw pointer issue would also happen without the lambdas (doing the `std::get<x>` in the constructor scope).

Reply:
I think we can resolve this?